test.cmake

```diff
@@ -109,8 +109,15 @@ ENDMACRO(ADD_UNIT_TEST NAME SOURCE)
 #   .. note:: :command:`FINDPYTHON` should have been called first.
 #
 MACRO(ADD_PYTHON_UNIT_TEST NAME SOURCE)
-  ADD_TEST(NAME ${NAME} COMMAND ${PYTHON_EXECUTABLE} "${PROJECT_SOURCE_DIR}/${SOURCE}")
-  SET(PYTHONPATH)
+  IF(ENABLE_COVERAGE)
+    SET(PYTHONPATH "${CMAKE_INSTALL_PREFIX}/${PYTHON_SITELIB}")
+    ADD_TEST(NAME ${NAME}
+      COMMAND ${PYTHON_EXECUTABLE} -m coverage run --branch -p --source=${PYTHONPATH} "${PROJECT_SOURCE_DIR}/${SOURCE}"
+      WORKING_DIRECTORY ${PROJECT_BINARY_DIR})
```

Comment:
This might be a subtle breaking change when coverage is enabled

Reply:
I already found some of those breaking changes in hpp-fcl, as a python helper script is available in the test source dir, but not in the installed project

Reply:
If you don't mind requiring CMake 3.7 (notably this excludes "stock" CMake in 16.04) you can use [test fixtures](https://cmake.org/cmake/help/latest/prop_test/FIXTURES_REQUIRED.html) to move the generated coverage files after the run. The cleanup test can be something like `${CMAKE_COMMAND} -DPROJECT_BINARY_DIR="${PROJECT_BINARY_DIR}" -DSOURCE=${SOURCE} -P cleanup-python-coverage.cmake` where `cleanup-python-coverage.cmake` is some CMake script that boils down to `mv ${SOURCE}.lcov ${PROJECT_BINARY_DIR}` (but with `${CMAKE_COMMAND} -E`)

If you want to keep backward compatibility you can use a regular dependency between the test and the cleanup test but you will lose the benefits of the fixture (e.g. running the test alone automatically runs the cleanup) but I think it's relatively rare to enable coverage outside of CI.

Reply:
I don't see why we would need to move the generated coverage files

Reply:
Assuming you want to get rid of this `WORKING_DIRERCTORY` then I guess the coverage file would be generated in the default working directory and you'd want to move them to a predictable place so the coverage report can be correctly generated.

But if you are fine with fixing breaking changes introduced by this I don't mind much since I don't use the macro myself.

Reply:
I see. I wasn't considering this `WORKING_DIRECTORY` to be an issue. For me the issue might come from the `--source=${PYTHONPATH}`, because we need to install the project before beeing able to get the coverage. 

The cause is that we need all compiled module to be available, so the source tree is not an option, and we also need the `.py` file, so the build tree is not an option (as, for now, we only have `.pyc` there), and that's why I tried with the install tree.

But in some cases, the install tree might lack python files defined in the source tree and not meant to be installed (eg. https://github.com/humanoid-path-planner/hpp-fcl/blob/devel/test/python_unit/test_case.py)
Maybe this is a design error, as one might expect the test to be available with only the install tree available, eg. after an install from a binary package, so I guess I'll try with this first.

About the changes because of the `WORKING_DIRECTORY`, you're probably right, but my ideas will be more clear once I get an actual error in the wild